src/corehost/cli/json_parser.cpp

```diff
@@ -0,0 +1,94 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "json_parser.h"
+#include "rapidjson/error/en.h"
+#include "utils.h"
+#include <cassert>
+#include <cstdint>
+
+void json_parser_t::realloc_buffer(size_t size)
+{
+    m_json.reset(new pal::char_t[size + 1]);
+    m_json[size] = '\0';
+}
+
+bool json_parser_t::parse_json(const pal::string_t& context)
+{
+    assert(m_json);
+
+    m_document.ParseInsitu(m_json.get());
+
+    if (m_document.HasParseError())
+    {
+        trace::error(_X("A JSON parsing exception occurred in [%s]: %s"),
+                     context.c_str(),
+                     rapidjson::GetParseError_En(m_document.GetParseError()));
+        return false;
+    }
+
+    if (!m_document.IsObject())
+    {
+        trace::error(_X("Expected a JSON object in [%s]"), context.c_str());
+        return false;
+    }
+
+    return true;
+}
+
+bool json_parser_t::parse_string(const pal::string_t& str, const pal::string_t& context)
+{
+    realloc_buffer(str.length());
+    memcpy(m_json.get(), str.c_str(), str.length());
```

Comment:
The code unit size should be considered whenever using low level C memory functions like `memcpy`.

``` c++
size_t byte_len = str.length() * sizeof(str[0]);
realloc_buffer(byte_len);
memcpy(m_json.get(), str.c_str(), byte_len);
```

Reply:
Argh. Yeah, adding this to the bag of stuff to worry about. Not really used to wide strings.

Reply:
Pushed a new version where this change won't b e necessary; data in m_json will be kept in UTF-8, and m_document will handle the UTF8 -> UTF-16 goo.